### PR TITLE
dev-libs/protobuf: version bump to 3.3.0

### PR DESCRIPTION
**Pull request by lluixhi**
Fix test phase in protobuf-3.x
Fix test in protobuf-2.6.1-protoc-cmdline.patch
Add compatibility patch for musl libc and freebsd. (from upstream)
Remove vim-syntax USE flag.
Fix Licenses (The LICENSE/COPYING.txt file has been a 3-clause BSD license since <2.5.0, so I'm not sure why it was set to Apache-2.0)
Enable C++ implementation in protobuf-python, fix python3 incompatibility in patch.

Gentoo-Bug: https://bugs.gentoo.org/460608
Gentoo-Bug: https://bugs.gentoo.org/544420
Gentoo-Bug: https://bugs.gentoo.org/602182
Gentoo-Bug: https://bugs.gentoo.org/611180

### CODE REVIEW

**Comment by lluixhi** [comment]
Bump

**Comment by lluixhi** [comment]
Bumping again..

**Comment by lluixhi** [comment]
@floppym @monsieurp

**Comment by monsieurp** [comment]
@lluixhi I've sent @vapier an email to let him know about this pending PR. As far as the Java and Python ebuilds are concerned, it looks good to me. Could you provide a build log for these parts though? So that we can make sure they do compile. Cheers!

**Comment by lluixhi** [comment]
@monsieurp turns out 3.2.1 was just released, so I'll probably bump the versions here and attach build.log's at 611180

**Comment by tomboy-64** [comment]
Thanks for getting the tests working again. I never gave that much attention as protobuf always seems to complain about missing .proto files from the testsuite. So even with the test-system working, ```FEATURES=test emerge protobuf``` will likely fail.

However, regarding your major-minor-patch, while it's a good thing that's fixed, it's highly problematic to include in a specific distribution. Proper procedure would be to get it accepted by upstream and propagated from there. For one, Gentoo users shouldn't be the only one profiting from that fix. On the other hand, if it breaks something within protobuf, have the original authors take a good look at it.

**TL;DR: keep the test-fix; send the major-minor-patch upstream.**

Other than that, the version bump looks good to me.

**Comment by lluixhi** [comment]
@tomboy-64 
Uh, that patch is two commits from upstream.
https://github.com/google/protobuf/commit/b7f25d42d3bd1b8cd9d995111e654a4601fe2ac8
https://github.com/google/protobuf/commit/416f90939d4de58fe1a4e2489120010313183291

**Comment by tomboy-64** [comment]
Ah, now that is a precious bit of info. So basically a backported fix.
I'll rebuild it tonight and do some compiling against it. See if anything explodes that I'm using.

**Comment by tomboy-64** [comment]
No issues for me. I'd suggest to merge it (including the major-minor patch).

**Comment by lluixhi** [comment]
Bump?

**Comment by mgorny** [comment]
@tomboy-64, I don't really understand the maintainer situation around this package. It lists you but not the proxy-maint project. Any idea why is it like that?

File: dev-libs/protobuf/protobuf-3.2.1.ebuild [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+multilib_src_test() {
+	emake check
+}
+
+multilib_src_install_all() {
+	if use vim-syntax; then
+		insinto /usr/share/vim/vimfiles/syntax
+		doins editors/proto.vim
+		insinto /usr/share/vim/vimfiles/ftdetect/
```
Comment (mgorny) [inline dev-libs/protobuf/protobuf-3.2.1.ebuild]: Nitpick: first insinto is not terminated by /, second is. But I'll fix this important issue when merging, don't worry.

File: dev-libs/protobuf/protobuf-3.2.1.ebuild [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+		elisp-compile "${S}"/editors/protobuf-mode.el
+	fi
+}
+
+multilib_src_test() {
+	emake check
+}
+
+multilib_src_install_all() {
+	if use vim-syntax; then
```
Comment (mgorny) [inline dev-libs/protobuf/protobuf-3.2.1.ebuild]: Per the rules we don't keep USE flags for 2 small files with no explicit deps. Again, I guess I'll just fix it for you to recompense for the delay ;-).

File: dev-python/protobuf-python/protobuf-python-3.2.1.ebuild [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+IUSE=""
+
+# Protobuf is only a build-time dep, but depends on the exact same version
+# (excluding revision), since we are using the same tarball.
+# In case of using the (linked) cpp implementation we should be fine with the same subslot.
+RDEPEND="${PYTHON_DEPS}
+	!<dev-libs/protobuf-3[python(-)]"
+
+DEPEND="${RDEPEND}
+	>=dev-libs/protobuf-3
```
Comment (mgorny) [inline dev-python/protobuf-python/protobuf-python-3.2.1.ebuild]: I think `:=` goes in here, unless it really doesn't link to the library.

Comment (lluixhi) [inline dev-python/protobuf-python/protobuf-python-3.2.1.ebuild]: It doesn't actually link to the library -- no shared objects included.

Comment (lluixhi) [inline dev-python/protobuf-python/protobuf-python-3.2.1.ebuild]: Hmmm. It looks like by default it is building without the c++ extension so it doesn't link to the library. Should this be changed..?

Comment (lluixhi) [inline dev-python/protobuf-python/protobuf-python-3.2.1.ebuild]: On second thought, yes -- I'm going to change this to build the c++ implementation by default.

**Comment by lluixhi** [comment]
@mgorny 
I addressed the changes that you suggested because I realized the python C++ extension was never getting built, and I wanted to enable it.

**Comment by mgorny** [comment]
I've got one failing test (for x86_32 ABI, using your previous protobuf ebuild but I suppose nothing changes):

```
[ RUN      ] EncodeDecodeTest.RedirectInputOutput
/tmp/portage/dev-libs/protobuf-3.2.1/work/protobuf-3.2.1/src/google/protobuf/compiler/command_line_interface_unittest.cc:2070: Failure
Value of: Run(cmd)
  Actual: false
Expected: true
/tmp/portage/dev-libs/protobuf-3.2.1/work/protobuf-3.2.1/src/google/protobuf/compiler/command_line_interface_unittest.cc:1990: Failure
Value of: File::ReadFileToString(actual_file, &actual_output)
  Actual: false
Expected: true
[  FAILED  ] EncodeDecodeTest.RedirectInputOutput (24 ms)
```

Any idea what that is? Also, does the Python counterpart have some tests? If it does, you should enable them.

File: dev-java/protobuf-java/protobuf-java-3.2.1.ebuild [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+LICENSE="BSD"
+SLOT="0/12"
+KEYWORDS="~amd64 ~arm ~arm64 ~ppc64 ~sh ~x86 ~amd64-linux ~arm-linux ~x86-linux ~x64-macos ~x86-macos"
+IUSE="nano"
+
+# Protobuf is only a build-time dep, but depends on the exact same version
+# (excluding revision), since we are using the same tarball.
+# But probably same subslot is sufficient.
+DEPEND=">=virtual/jdk-1.7
+	>=dev-libs/protobuf-3"
```
Comment (chewi) [inline dev-java/protobuf-java/protobuf-java-3.2.1.ebuild]: I've been asked whether this dependency should be stricter. There doesn't seem to be any native code involved but this [README](https://github.com/google/protobuf/blob/master/java/README.md) says you should use the same version. I would change it to `~dev-libs/protobuf-${PV}`. They should always get bumped in tandem anyway.

**Comment by lluixhi** [comment]
@mgorny 
Any idea if those are tests broken by the protobuf-2.6.1-protoc-cmdline patch?
I vaguely remember that that broke a couple of tests, but I think @vapier would know.

@chewi 
Yeah, sure.

**Comment by mgorny** [comment]
> Any idea if those are tests broken by the protobuf-2.6.1-protoc-cmdline patch?

It's possible. Comment it out and try for yourself. If it's that, then someone should eventually either fix those tests, or at least disable them. For now, though, I'd be happy enough with you confirming that it's not a 'real' issue.

**Comment by lluixhi** [comment]
Double-checked protobuf-python, all the unit tests are running (turns out tests failed on python3.6, so I removed it from the list of `PYTHON_TARGETS`.)

Also 3.3.0 was just released a couple days ago -- should I add that to this PR as well?

@mgorny, that test is actually included in the `protobuf-2.6.1-protoc-cmdline.patch`, I'm wondering how long it has been broken.

**Comment by lluixhi** [comment]
Hmm on second thought the two protobuf-python tests that fail with python3.6 fail because the error output changed -- should we worry about this?

```
======================================================================
FAIL: testInvalidTimestamp (google.protobuf.internal.json_format_test.JsonFormatTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/adb/Downloads/protobuf-3.3.0/python/google/protobuf/json_format.py", line 496, in _ConvertFieldValuePair
    self.ConvertMessage(value, sub_message)
ValueError: year 0 is out of range

During handling of the above exception, another exception occurred:

google.protobuf.json_format.ParseError: Failed to parse value field: year 0 is out of range.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/adb/Downloads/protobuf-3.3.0/python/google/protobuf/internal/json_format_test.py", line 772, in testInvalidTimestamp
    json_format.Parse, text, message)
AssertionError: "Failed to parse value field: year is out of range." does not match "Failed to parse value field: year 0 is out of range."

======================================================================
FAIL: testInvalidTimestamp (google.protobuf.internal.well_known_types_test.TimeUtilTest)
----------------------------------------------------------------------
ValueError: year 0 is out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/adb/Downloads/protobuf-3.3.0/python/google/protobuf/internal/well_known_types_test.py", line 289, in testInvalidTimestamp
    '0000-01-01T00:00:00Z')
AssertionError: "year is out of range" does not match "year 0 is out of range"

----------------------------------------------------------------------
```

File: dev-python/protobuf-python/protobuf-python-3.2.1.ebuild [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+SLOT="0/12"
+KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~ia64 ~mips ~ppc ~ppc64 ~sh ~sparc ~x86 ~amd64-linux ~arm-linux ~x86-linux ~x64-macos ~x86-macos"
+IUSE=""
+
+# Protobuf is only a build-time dep, but depends on the exact same version
+# (excluding revision), since we are using the same tarball.
+# In case of using the (linked) cpp implementation we should be fine with the same subslot.
+RDEPEND="${PYTHON_DEPS}
+	!<dev-libs/protobuf-3[python(-)]
+	~dev-libs/protobuf-${PV}"
```
Comment (lluixhi) [inline dev-python/protobuf-python/protobuf-python-3.2.1.ebuild]: @mgorny 
Not sure, will portage rebuild this package on a version bump of `dev-libs/protobuf`, similar to `:=`?

Comment (mgorny) [inline dev-python/protobuf-python/protobuf-python-3.2.1.ebuild]: Good question. If it's not triggered for upgrade, Portage might just decide not to upgrade protobuf instead. `:=` is better for that, if you can trust that they use only the public API.

**Comment by mgorny** [comment]
> Also 3.3.0 was just released a couple days ago -- can I add that to this PR, or should I open a new one?

Preferably just replace 3.2.1 with 3.3.0, unless you have a good reason to keep both versions around.

**Comment by lluixhi** [comment]
Hmm. SONAME changes between versions, and there appear to be some (minimal) incompatibilities.

Under C++:
> Parser now rejects field number 0 correctly.
Under java:
> Map field setters eagerly validate inputs and throw NullPointerExceptions as appropriate.
> proto3 enum's getNumber() method now throws on UNRECOGNIZED values.
> Output of JsonFormat is now locale independent.

I think I'll add 3.3.0 as well.

**Comment by mgorny** [comment]
Actually, I wouldn't consider that a good reason because:

1. People would upgrade straight to 3.3.0 and 3.2.1 will receive barely any real testing,
2. someone will soon want to remove 3.2.1 as unused redundant version,
3. but he will unnecessarily be worried that someone might be relying on this specific version without leaving any trace of that.

So, just don't add unnecessary old versions. That's only increases workload on people cleaning them up afterwards.

**Comment by lluixhi** [comment]
@mgorny 
Alright, I'll remove 3.2.1, and this can hopefully finally get merged :P

**Comment by tomboy-64** [comment]
@mgorny It lists me as maintainer because I was the only one who cared for a protobuf-3.0-beta version of the package in the tree. Why proxy-maint is not listed I do not know. Probably nobody had a close enough look at things back when I pushed things back then.

Also, protobuf-python should include python3.6 in PYTHON_COMPAT. I just checked 3.3.0, compiles fine and simple test worked as well.

**Comment by tomboy-64** [comment]
@lluixhi Please also add in the commit message for protobuf-python that you switched to the C++ implementation! This is not a trivial version bump right there.

**Comment by lluixhi** [comment]
@tomboy-64 
I already mentioned that the C++ implementation is enabled in the commit message, unless you think I should put it in the first line?

Also, I initially enabled python3_6 support, but there are a couple of broken tests, at least on my end:
https://github.com/google/protobuf/issues/3037

**Comment by mgorny** [comment]
@chewi, is java happy already? If so, I'll merge it.

**Comment by chewi** [comment]
Go for it.

**Comment by tomboy-64** [comment]
Sorry for the last minute intrusion.

I reviewed the python stuff again and my stomach hurts with the switch to the cpp-implementation of python-protobuf. I simply didn't notice the extensive changes made to protobuf-python - i previously presumed it was a simple version bump.
The cpp implementation used to have a note about being buggy; it has vastly improved but obviously it's still less robust than the native python implementation.

This is the change I consider necessary:
https://github.com/gentoo/gentoo/commit/900ecf0aef4cddf3f5b0ab79a18b6195d96c96de
Combined with lluixhi's changes:
https://github.com/gentoo/gentoo/compare/master...tomboy-64:protobuf-python-3.3.0?expand=1
As bare diff:
```
diff --git a/dev-python/protobuf-python/protobuf-python-3.3.0.ebuild b/dev-python/protobuf-python/protobuf-python-3.3.0.ebuild
index 413926d1f4..1c20ef3873 100644
--- a/dev-python/protobuf-python/protobuf-python-3.3.0.ebuild
+++ b/dev-python/protobuf-python/protobuf-python-3.3.0.ebuild
@@ -3,7 +3,7 @@

 EAPI=6
 # pypy fails tests; pypy3 fails even running tests
-PYTHON_COMPAT=( python2_7 python3_{4,5} )
+PYTHON_COMPAT=( python2_7 python3_{4,5,6} )

 inherit distutils-r1

@@ -17,7 +17,7 @@ SRC_URI="https://github.com/google/protobuf/archive/v${MY_PV}.tar.gz -> protobuf
 LICENSE="BSD"
 SLOT="0/13"
 KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~ia64 ~mips ~ppc ~ppc64 ~sh ~sparc ~x86 ~amd64-linux ~arm-linux ~x86-linux ~x64-macos ~x86-macos"
-IUSE=""
+IUSE="-cxx"

 # Protobuf is only a build-time dep, but depends on the exact same version
 # (excluding revision), since we are using the same tarball.
@@ -34,9 +34,23 @@ PATCHES=( "${FILESDIR}/${PN}-3.0.0_beta3-link-against-installed-lib.patch" )

 S="${WORKDIR}/protobuf-${MY_PV}/python"
 python_compile() {
-       distutils-r1_python_compile --cpp_implementation
+       if use cxx ; then
+               distutils-r1_python_compile --cpp_implementation
+       else
+               default
+       fi
 }

 python_test() {
+       use !cxx && distutils_install_for_testing
        esetup.py test
 }
+
+pkg_postinst() {
+       if use cxx ; then
+               elog "You chose to install the C++-implementation of protobuf-python."
+               elog "The C++ implementation is newer, more performant and less robust than the native"
+               elog "Python implementation. If you encounter problems, you are advised to first try"
+               elog "dev-python/protobuf-python[-cxx] before seeking help."
+       fi
+}
```

**Comment by mgorny** [comment]
I'm just going to keep the Python as it was before. Then fix the namespace. Then we can wonder about extra implementations.
